cluster/manager.go

```diff
@@ -0,0 +1,229 @@
+		return
+	}
+	peer := Node{}
+	err := json.Unmarshal(node.Meta, &peer)
+	if err != nil {
+		panic(err)
+	}
+	peer.RemoteAddr = node.Addr.String()
+	if peer.Name == c.node.Name {
+		peer.local = true
+	}
+	c.Peers[node.Name] = peer
+	log.Info("Node %s at %s has been updated - %s\n", node.Name, node.Addr.String(), node.Meta)
+}
+
+func (c *ClusterManager) BroadcastUpdate() {
+	if c.list != nil {
+		go c.list.UpdateNode(time.Second)
+	}
+}
+
+// NodeMeta is used to retrieve meta-data about the current node
+// when broadcasting an alive message. It's length is limited to
+// the given byte size. This metadata is available in the Node structure.
+func (c *ClusterManager) NodeMeta(limit int) []byte {
+	c.Lock()
+	meta, err := json.Marshal(c.node)
+	c.Unlock()
+	if err != nil {
+		log.Fatal(4, err.Error())
+	}
+	return meta
+}
+
+// NotifyMsg is called when a user-data message is received.
+// Care should be taken that this method does not block, since doing
+// so would block the entire UDP packet receive loop. Additionally, the byte
+// slice may be modified after the call returns, so it should be copied if needed.
+func (c *ClusterManager) NotifyMsg(buf []byte) {
+	// we dont have any need for passing messages between nodes, other then
+	// the NodeMeta sent with alive messages.
+	return
+}
+
+// GetBroadcasts is called when user data messages can be broadcast.
+// It can return a list of buffers to send. Each buffer should assume an
+// overhead as provided with a limit on the total byte size allowed.
+// The total byte size of the resulting data to send must not exceed
+// the limit. Care should be taken that this method does not block,
+// since doing so would block the entire UDP packet receive loop.
+func (c *ClusterManager) GetBroadcasts(overhead, limit int) [][]byte {
+	// we dont have any need for passing messages between nodes, other then
+	// the NodeMeta sent with alive messages.
+	return nil
+}
+
+// LocalState is used for a TCP Push/Pull. This is sent to
+// the remote side in addition to the membership information. Any
+// data can be sent here. See MergeRemoteState as well. The `join`
+// boolean indicates this is for a join instead of a push/pull.
+func (c *ClusterManager) LocalState(join bool) []byte {
+	c.Lock()
+	c.Peers[c.node.Name] = c.node
+	meta, err := json.Marshal(c.Peers)
+	c.Unlock()
+	if err != nil {
+		log.Fatal(4, err.Error())
+	}
+	return meta
+}
+
+func (c *ClusterManager) MergeRemoteState(buf []byte, join bool) {
+	knownPeers := make(map[string]Node)
+	err := json.Unmarshal(buf, &knownPeers)
+	if err != nil {
+		log.Fatal(4, err.Error())
+	}
+	c.Lock()
+	for name, meta := range knownPeers {
+		if existing, ok := c.Peers[name]; ok {
+			if meta.Updated.After(existing.Updated) {
+				log.Info("updated node meta found in state update for %s", meta.Name)
+				c.Peers[name] = meta
+			}
+		} else {
+			log.Info("new node found in state update. %s", meta.Name)
+			c.Peers[name] = meta
+		}
+	}
+	c.Unlock()
```

Comment:
again, i think it would be more appropriate to put `defer c.Unlock()` right after `c.Lock()` because it's very easy for someone to add a return inside one of the enclosed conditions and missing the lock.